Added OTLP Attributes and testing

leggebroten opened the pull request:
Since CQRS is considered a messaging system, changed emitted attributes to follow that semantic and added tests.

Assuming the testing of tracing events is adequate, all that's needed is to assert that we've attached correctly to the telemetry callbacks and that the spans are correctly constructed in the three cases: success, error, exception


closes: #1 , #2 , #3 , #4

leggebroten committed:
Beginnings of a test framework

leggebroten committed:
Tests for Aggregate and updated OTLP attributes

leggebroten committed:
Tests for Application and updated OTLP attributes

leggebroten committed:
Tests for Event Handler and updated OTLP attributes

leggebroten committed:
Tests for Process Manager and updated OTLP attributes

leggebroten committed:
Event Store implementation and tests

* Changed mix dependency for Commanded to get unreleased merge DavyDog made for event store
* Changed callback module to capture correct set of events
* Modified attributes to match messaging semantics like others w/i this library

leggebroten commented on lib/commanded/event_handler.ex:
```diff
@@ -34,22 +34,28 @@ defmodule OpentelemetryCommanded.EventHandler do
   def handle_start(_event, _measurements, meta, _) do
     event = meta.recorded_event
+
+    # TODO: do we want to trap this with helpful message about needing Middleware?
```
@davydog187 and/or @bryannaegele 
If the user hasn't added the middleware to their app this will cause an exception and kills this process.

Should I trap here?

davydog187 replied:
My understanding of the design philosophy behind Otel is to fail gracefully, i.e. it shouldn't impact the uptime of the system. So catching/rescuing any issues here and downgrading the system with a helpful error or warning log seems like a good move

leggebroten commented on lib/commanded/event_store.ex:
```diff
@@ -3,48 +3,113 @@ defmodule OpentelemetryCommanded.EventStore do
+      "messaging.system": "commanded",
+      "messaging.protocol": "cqrs",
+      "messaging.destination_kind": "event_store",
+      # TODO does this need to change based on the event?  Or maybe `internal`?
+      "messaging.operation": "receive",
```
@bryannaegele 
For spans created by a messaging system all have operation types from that pool (send, receive, subscribe)?

This doesn't seem to fit well.  It seems an "internal" span w/i the CQRS messaging system.

Thoughts?

bryannaegele replied:
I don't know where `subscribe` comes from. Messaging operations are `send, receive, process`. Span type of `internal` is only appropriate where a function is calling another function. If there's a message or operation crossing a system boundary, e.g. to the database, then it's not internal.

CQRS is a bit of a hybrid between Messaging and RPC, so I'm not sure of all the specifics of how it should be treated. You can always solicit opinions in the `#opentelemetry` channel of the CNCF slack.

leggebroten commented on lib/commanded/event_store.ex:
```diff
@@ -3,48 +3,113 @@ defmodule OpentelemetryCommanded.EventStore do
+      "messaging.operation": "receive",
+      "messaging.message_id": event.causation_id,
+      "messaging.conversation_id": event.correlation_id,
+#      "messaging.destination": meta.handler_module,
```
@davydog187 and/or @bryannaegele 
The above question about the span kind being "internal" is supported by the fact there is no known source or destination for this span.

leggebroten commented on lib/commanded/process_manager.ex:
```diff
@@ -38,17 +38,25 @@ defmodule OpentelemetryCommanded.ProcessManager do
+      "messaging.system": "commanded",
+      "messaging.protocol": "cqrs",
+      "messaging.destination_kind": "process_manager",
+      "messaging.operation": "receive",
```
@bryannaegele 
A Process Manager "receives" events and "sends" commands.
Is this correct?

leggebroten replied:
@bryannaegele 
Sort of.  It `receive`s events, and its `handle` call may return Commands.
The Process Manager does not `send` any Commands it produces instead Commanded hands them back to the app to be dispatched

leggebroten commented on lib/commanded/process_manager.ex:
```diff
@@ -38,17 +38,25 @@ defmodule OpentelemetryCommanded.ProcessManager do
+      "messaging.commanded.stream_id": event.stream_id,
+      "messaging.commanded.stream_version": event.stream_version,
+      "messaging.commanded.handler_name": meta.process_manager_name
+      # TODO add back
+      # consistency: meta.consistency,
+      #  TODO add this back into commanded
+      # "event.last_seen": meta.last_seen_event
```
Left @davydog187 's comments.

leggebroten replied:
@bryannaegele 
After reading a ton, I think that I should remove the `messaging.` prefix from the Commanded attributes.
This would also allow easier query building too I would think
Thoughts?

e.g. 
```
      "messaging.commanded.application": meta.application,
      "messaging.commanded.event": recorded_event.event_type,
      "messaging.commanded.event_id": recorded_event.event_id,
      "messaging.commanded.event_number": recorded_event.event_number,
      "messaging.commanded.process_uuid": meta.process_uuid,
      "messaging.commanded.stream_id": recorded_event.stream_id,
      "messaging.commanded.stream_version": recorded_event.stream_version,
      "messaging.commanded.handler_name": meta.process_manager_name
```

should be
```
      "commanded.application": meta.application,
      "commanded.event": recorded_event.event_type,
      "commanded.event_id": recorded_event.event_id,
      "commanded.event_number": recorded_event.event_number,
      "commanded.process_uuid": meta.process_uuid,
      "commanded.stream_id": recorded_event.stream_id,
      "commanded.stream_version": recorded_event.stream_version,
      "commanded.handler_name": meta.process_manager_name
```

davydog187 commented on lib/commanded/aggregate.ex:
```diff
@@ -38,15 +38,18 @@ defmodule OpentelemetryCommanded.Aggregate do
-      application: meta.application,
-      "causation.id": context.causation_id,
-      "correlation.id": context.correlation_id,
-      "aggregate.function": context.function,
```
do we still want these two fields?

leggebroten replied:
`aggregate.function": context.function` is still there, just massaged to fit naming for messaging attributes:
```elixir
      "messaging.commanded.function": context.function
```

I can restore `context.lifespan` but while testing, didn't see it's value as an attribute since it doesn't effect the processing of the current span.  Thoughts?

leggebroten committed:
Found bug where context wasn't always being propagated across process boundaries.
Fixed and added tests.

Safely handle case where the user hasn't added middleware `middleware(OpentelemetryCommanded.Middleware)` to the router

bryannaegele commented on lib/commanded/event_store.ex:
```diff
@@ -3,48 +3,111 @@ defmodule OpentelemetryCommanded.EventStore do
+
+    attributes = [
+      "messaging.system": "commanded",
+      "messaging.protocol": "cqrs",
```
`cqrs` isn't a transport protocol, it's a pattern. This attribute should probably not be included.

leggebroten committed:
Best guess so far with attributes.

leggebroten committed:
Missed two

davydog187 reviewed (approved):
Appreciate the work here, @leggebroten
